api/src/main/java/org/apache/iceberg/expressions/ResidualEvaluator.java

```diff
@@ -211,31 +210,48 @@ public Expression or(Expression leftResult, Expression rightResult) {
         return pred; // not associated inclusive a partition field, can't be evaluated
       }
 
-      List<UnboundPredicate<?>> strictProjections = Lists.transform(parts,
-          part -> ((Transform<T, ?>) part.transform()).projectStrict(part.name(), pred));
+      for (PartitionField part : parts) {
 
-      if (Iterables.all(strictProjections, Objects::isNull)) {
-        // if there are no strict projections, the predicate must be in the residual
-        return pred;
-      }
+        // checking the strict projection
+        UnboundPredicate<?> strictProjection = ((Transform<T, ?>) part.transform()).projectStrict(part.name(), pred);
+        Expression strictResult = null;
+
+        if (strictProjection != null) {
+          Expression bound = strictProjection.bind(spec.partitionType(), caseSensitive);
+          if (bound instanceof BoundPredicate) {
+            strictResult = super.predicate((BoundPredicate<?>) bound);
+          } else {
+            strictResult = bound;
+          }
+        }
+
+        if (strictResult != null && strictResult.equals(Expressions.alwaysTrue())) {
```

Comment:
Checking the expression op is probably better: `strictResult.op() == Expression.Operation.TRUE`

Reply:
Ack.